src/config/config.ts

```diff
@@ -184,68 +149,70 @@ export class Config extends ConfigFile<ConfigFile.Options> {
-        {
-          key: 'restDeploy',
-          hidden: true,
-          input: {
-            validator: value => value != null && ['true', 'false'].includes(value.toString()),
-            failedMessage: Config.messages.getMessage('InvalidBooleanConfigValue')
-          }
-        },
-        {
-          key: Config.MAX_QUERY_LIMIT,
-          input: {
-            validator: value => isNumber(value),
-            failedMessage: Config.messages.getMessage('InvalidNumberConfigValue')
-          }
+    Config.allowedProperties = [
+      {
+        key: 'instanceUrl',
+        input: {
+          // If a value is provided validate it otherwise no value is unset.
+          validator: value => value == null || (isString(value) && sfdc.isSalesforceDomain(value)),
+          failedMessage: Config.messages.getMessage('InvalidInstanceUrl')
         }
-      ];
+      },
+      {
+        key: Config.API_VERSION,
+        hidden: true,
+        input: {
+          // If a value is provided validate it otherwise no value is unset.
+          validator: value => value == null || (isString(value) && sfdc.validateApiVersion(value)),
+          failedMessage: Config.messages.getMessage('InvalidApiVersion')
+        }
+      },
+      { key: Config.DEFAULT_DEV_HUB_USERNAME },
+      { key: Config.DEFAULT_USERNAME },
+      {
+        key: Config.ISV_DEBUGGER_SID,
+        encrypted: true,
+        input: {
+          // If a value is provided validate it otherwise no value is unset.
+          validator: value => value == null || isString(value),
+          failedMessage: Config.messages.getMessage('InvalidIsvDebuggerSid')
+        }
+      },
+      {
+        key: Config.ISV_DEBUGGER_URL,
+        input: {
+          // If a value is provided validate it otherwise no value is unset.
+          validator: value => value == null || isString(value),
+          failedMessage: Config.messages.getMessage('InvalidIsvDebuggerUrl')
+        }
+      },
+      {
+        key: Config.DISABLE_TELEMETRY,
+        input: {
+          validator: value => value == null || ['true', 'false'].includes(value.toString()),
+          failedMessage: Config.messages.getMessage('InvalidBooleanConfigValue')
+        }
+      },
+      // This should be brought in by a plugin, but there isn't a way to do that right now.
+      {
+        key: 'restDeploy',
+        hidden: true,
+        input: {
+          validator: value => value != null && ['true', 'false'].includes(value.toString()),
+          failedMessage: Config.messages.getMessage('InvalidBooleanConfigValue')
+        }
+      },
+      {
+        key: Config.MAX_QUERY_LIMIT,
+        input: {
+          validator: value => isNumber(value),
+          failedMessage: Config.messages.getMessage('InvalidNumberConfigValue')
+        }
+      }
+    ];
+
+    if (options && options.customConfigProperties) {
+      Config.allowedProperties.push(...options.customConfigProperties);
```

Comment:
I think we might want a global method to do this rather than in the constructor. That method might have to protect against plugin versions too... we will have to guard against version differences but I think we will run into more problems without it.